### ISSUE TITLE
[Cloudinary] Image URL in Editor preview broken with output_filename_only option 

### ISSUE THREAD

**Issue by ksherman** [netlify/netlify-cms#1934]
**Describe the bug**
When using the Cloudinary Media Library integration, I want to use the `output_filename_only` option to take advantage of URL transformations. However, this option seems to break the rendering of the preview images.

**To Reproduce**
Switch Cloudinary image assets to `output_filename_only` and observer the broken image links in the editor.

**Expected behavior**
We would expect the editor preview components to build the preview URL for the image to append the the image filename.

**Screenshots**
![screenshot 2018-12-04 22 05 50](https://user-images.githubusercontent.com/2822402/49489550-2db88500-f811-11e8-9c03-7a480995233a.png)

**Applicable Versions:**
Netlify CMS 2.3.0

**CMS configuration**
https://gist.github.com/ksherman/cb5b319858b8ea06fbf638818f839d84.js

**Additional context**
You all are awesome and I know this feature was juuuuust released :D

**Comment by kishaningithub** [comment]
One additional point here, is i can organize images into folders in cloudinary at which point the output filename should be `<folder name>/<file name>` instead of just `filename`

**Comment by danlutz** [comment]
Awesome work on the Cloudinary integration!

I think what is needed here, is that the CMS integrates the "cloudinary-react" package from npm, gets the cloudname from the config.yml, and then displays the React component like this:

```js
import React from 'react'
import { Image, Transformation } from 'cloudinary-react'

const PreviewImage = ({imgPublicId, cloudName, appropriateWidth}) => 
    (<Image publicId={imgPublicId} cloudName={cloudName}>
          <Transformation
            width="appropriateWidth"
            crop="fill"
            dpr="auto"
            fetchFormat="auto"
            quality="auto"
          />
    </Image>)

```

I'll take a look and see if I can make a PR, but don't expect too much

**Comment by danlutz** [comment]
Also, if you want to inline an image inside the "body", it only passes the filename to the "src" attribute.

It tried digging through the source code, but unfortunatly I could not find the correct place to code this additional logic. Looking forward to someone that is more skilled than me 💻

**Comment by erquhart** [comment]
I'll be going through PR's soon, but I think the best way to handle this is probably by some cooperation between the media library and the image widget preview component. Still need to think through how that would work, but the logic for connecting the url value with the Cloudinary root url is simple, just a matter of where that should occur and whether it's possible without requiring configuration. I think it is.

**Comment by hringleikur** [comment]
I agree that the thumbnail at least should be possible with zero config. The main issue I see is the preview - either it would need to be left broken or we would have to make a design decision that the user might disagree with.

**Comment by erquhart** [comment]
Widget preview components provide a default preview for an individual field. We would make the image widget provide the image as expected by default. If someone _wants_ the actual field value, they can still access it in a custom preview template (or custom widget preview component), so everyone wins.

The challenge is doing this in a generic way so that the image widget makes no assumptions about your media library, hence the need to provide some sort of cooperative API between these two extension points.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by aamorozov** [comment]
The issue seems be related to the image widget in general - i'm using manual upload of the images and they show up as broken links in the admin, although the frontend app is still able to query them.

**Comment by erquhart** [comment]
Yep, those are `img` elements, if you check their `src` you'll see what they're looking for and can determine why it's not there. Either the image wasn't deployed, your CMS is not on the same domain as the site where the image is deployed, you're developing locally, etc. Pulling through the API instead of just constructing URLs is the solution.

**Comment by theoephraim** [comment]
+1 - this is pretty much required for the `output_filename_only` option to be usable at all.

For now I'll be parsing the full URL and inserting my transformations that way, but it would be great if Netlify CMS was able to realize that I'm using an image hosting service capable of resizing and use thumbnails instead of full rez images in the admin interface.

**Comment by emiled** [comment]
I encountered this issue also - the workaround I found was to set `output_filename_only: false`, and in the html templates, use a [Hugo snippet](https://aaronlauterer.com/blog/2017/02/hugo-get-last-and-first-element-from-url/) to split the string and recover the asset's name, isolated from the full path. 

With this asset name string, I appended it to the Cloudinary transform URL string that was set to a variable.


1. set the cloudinaryAssetName variable using Scratch (replace ".cloudinaryFullURL" with your path)

`        {{ $.Scratch.Set "cloudinaryAssetName"  (index (split .cloudinaryFullURL "/") (sub (len (split .cloudinaryFullURL "/")) 1)) }}
`

2. set the Cloudinary transform path to a variable, I did this inside the "/content/_index.md" frontmatter 

`cloudinaryPathThumb: "https://res.cloudinary.com/userid/image/upload/w_200,c_scale/"
`


3. Assign the image path with the concatenated URL -
`<img src='{{ $.Params.cloudinaryPathThumb }}{{ $.Scratch.Get "cloudinaryAssetName" }}'`


This outputs the desired path: 
`<img src='https://res.cloudinary.com/userid/image/upload/w_200,c_scale/assetname.jpg'`

**Comment by pavelbinar** [comment]
Currently there is still no solution for the option

`output_filename_only: true`

I would like to keep this open @erquhart

**Comment by jeromecoupe** [comment]
Subscribing to this. When using RWD images heavily, it's a lot easier to use Cloudinary URLs in templates to do the heavy lifting. The preview for the image is broken in the CMS when using

`output_filename_only: true` 

I understand the ideal would be a zero config solution, but wouldn't the most flexible option still be to add an `output_preview_path` ? 

- would allow admin to specify a transformed image
- would be independent from the used service

**Comment by mark-making** [comment]
> Currently there is still no solution for the option
> 
> `output_filename_only: true`
> 
> I would like to keep this open @erquhart

Such a great implementation, it would be great for the image preview to work in the editor. 😎

**Comment by ankitshekhawat** [comment]
> 1. Set the cloudinaryAssetName variable using Scratch (replace ".cloudinaryFullURL" with your path variable)
> 
> ` {{ $.Scratch.Set "cloudinaryAssetName" (index (split .cloudinaryFullURL "/") (sub (len (split .cloudinaryFullURL "/")) 1)) }}`
> 
> 1. Assign the image path with the concatenated URL -
>    `<img src='{{ $.Params.cloudinaryPathThumb }}{{ $.Scratch.Get "cloudinaryAssetName" }}'`

You could also use Go's built-in parse function:
`<img src='{{ $.Params.cloudinaryPathThumb }}{{ path.Base (urls.Parse .cloudinaryFullURL ) }}' />`

Maybe write this in a shortcode to get to check if the Host is Cloudinary to conditionally parse the image

**Comment by redraw** [comment]
Bumping this old thread. 
The problem is that if storing the full path, full resolution images get loaded in the admin as thumbnails, that's crazy...

Still, we're not able to set `output_filename_only: true` along with some `preview_path` field to build a transformation url for the admin thumbnails.

Any ideas?

**Comment by redraw** [comment]
Ok, I've did the following if you are in the same situation:

1. set `output_filename_only: false`
2. add a default transformation at top level
```
media_library:
  name: cloudinary
  config:
    cloud_name: ...
    api_key: ...
    default_transformations:
      - - fetch_format: auto
          width: 160
          quality: auto
          crop: scale
```
3. that saved the image with the default transformation in the URL, then in your code replace image/upload/(.+)/ with your desired transformation
